libgnucash/app-utils/gnc-option-date.hpp

```diff
@@ -0,0 +1,180 @@
+*/
+#ifndef GNC_OPTION_DATE_HPP_
+#define GNC_OPTION_DATE_HPP_
+
+extern "C"
+{
+#include <gnc-date.h>
+}
+
+#include <vector>
+#include <iostream>
+/**
+ * Reporting periods relative to the current date.
+ *
+ * The original design allowed custom RelativeDatePeriods, but that facility is
+ * unused so we'll go with compiled-in enums.
+ */
+enum class RelativeDatePeriod : int
+{
+    ABSOLUTE = -1,
+    TODAY,
+    ONE_WEEK_AGO,
+    ONE_WEEK_AHEAD,
+    ONE_MONTH_AGO,
+    ONE_MONTH_AHEAD,
+    THREE_MONTHS_AGO,
+    THREE_MONTHS_AHEAD,
+    SIX_MONTHS_AGO,
+    SIX_MONTHS_AHEAD,
+    ONE_YEAR_AGO,
+    ONE_YEAR_AHEAD,
+    START_THIS_MONTH,
+    END_THIS_MONTH,
+    START_PREV_MONTH,
+    END_PREV_MONTH,
+    START_NEXT_MONTH,
+    END_NEXT_MONTH,
+    START_CURRENT_QUARTER,
+    END_CURRENT_QUARTER,
+    START_PREV_QUARTER,
+    END_PREV_QUARTER,
+    START_NEXT_QUARTER,
+    END_NEXT_QUARTER,
+    START_CAL_YEAR,
+    END_CAL_YEAR,
+    START_PREV_YEAR,
+    END_PREV_YEAR,
+    START_NEXT_YEAR,
+    END_NEXT_YEAR,
+    START_ACCOUNTING_PERIOD,
+    END_ACCOUNTING_PERIOD,
+};
+
+constexpr unsigned relative_date_periods =
+    static_cast<unsigned>(RelativeDatePeriod::END_ACCOUNTING_PERIOD) + 2;
+
+using RelativeDatePeriodVec = std::vector<RelativeDatePeriod>;
+
+/**
+ * Report whether the relative date represents a period offset to today's date
+ * rather than the beginning or end of a date range. For example ONE_MONTH_AGO
+ * will be made concrete as the same day as today in the previous month.
+ *
+ * @param period The Relative Date Period to check.
+ * @return true if the date is stand-alone.
+ */
+bool gnc_relative_date_is_single(RelativeDatePeriod);
+
+/**
+ * Report whether the relative date represents the beginning of a date
+ * range. For example START_LAST_MONTH is the beginning of a range.
+ *
+ * @param period The Relative Date Period to check.
+ * @return true if the date is the beginning of a date range
+ */
+bool gnc_relative_date_is_starting(RelativeDatePeriod);
+
+/**
+ * Report whether the relative date represents the end of a date range. For
+ * example END_LAST_MONTH is the end of a range.
+ *
+ * @param period The Relative Date Period to check.
+ * @return true if the date is the end of a date range.
+ */
+bool gnc_relative_date_is_ending(RelativeDatePeriod);
+
+/**
+ * Provide the string representation of a relative date for persisting the
+ * value. This string is not localizable.
+ *
+ * @param period The relative date period.
+ * @return A constant string or nullptr if the period is ABSOLUTE. The string's
+ * lifetime will be that of the Relative Date Period. It must not be freed and
+ * should be copied if the period might be destroyed before the using code is
+ * finished.
+ */
+const char* gnc_relative_date_storage_string(RelativeDatePeriod);
+
+/**
+ * Provide the string representation of a relative date for displaying
+ * value to a user. This string is not localizable.
```

Comment:
Separating storage name from display name is a very good improvement. Though as I understand it the display string is localizable (likely a copy-paste left-over).

Reply:
Darn, I even remember thinking that I needed to change that comment after pasting it.

As it happens though I think that the user-facing representation is obsolete because it was used in the multiselect option value docstrings that @christopherlam and @Bob-IT removed last spring.

Reply:
> Darn, I even remember thinking that I needed to change that comment after pasting it.
> 
> As it happens though I think that the user-facing representation is obsolete because it was used in the multiselect option value docstrings that @christopherlam and @Bob-IT removed last spring.

The comment with the typo is for `gnc_relative_date_display_string` which returns the name of the option to display to the user. I believe we still need that. You are probably referring to `gnc_relative_date_description` instead, which returns a longer description of the option and which I agree is probably obsolete now.